FF102 User define byte stream interface - not experimental (#16818)
mdn/content: files/en-us/web/api/streams_api/using_readable_byte_streams/index.md

They are intended for use cases where data might be supplied or requested in arbitrary sized and potentially very large chunks, and hence where avoiding making copies is likely to improve efficiency.

This article explains how readable byte streams compare to normal "default" streams, and how you create and consume them.

> **Note:** Readable byte streams are almost identical to "normal" readable streams and almost all of the concepts are the same.
> This article assumes that you already understand those concepts and will only be covering them superficially (if at all).
> If you're not familiar with the relevent concepts, please first read: [Using readable streams](/en-US/docs/Web/API/Streams_API/Using_readable_streams), [Streams concepts and usage overview](/en-US/docs/Web/API/Streams_API#concepts_and_usage), and [Streams API concepts](/en-US/docs/Web/API/Streams_API/Concepts).

## Overview

> Readable streams provides a consistent interface for streaming data from some underlying source, such as a file or socket, to a consumer, such as a reader, transform stream or writeable stream.

Author: ```suggestion
Readable streams provides a consistent interface for streaming data from some underlying source, such as a file or socket, to a consumer, such as a reader, transform stream or writable stream.
```

The main difference between {{domxref("ReadableByteStreamController")}} and the default controller ({{domxref("ReadableStreamDefaultController")}}) is that it has an additional property {{domxref("ReadableByteStreamController.byobRequest")}} of type {{domxref("ReadableStreamBYOBRequest")}}.
This represents a pending read request by a consumer that will be made as a zero-copy transfer from the underlying source.
The property will be `null` if there is no pending request.

A `byobRequest` is only made available when a read request is made on a readable byte stream and there is no data in the stream's internal queues (if there is data then the request is satisfied from those queues).

An underlying byte source that needs to transfer data must check the `byobRequest` property and, if it is available, use it to transfer data.
If the property is `null`, incoming data should instead be added to the stream's internal queues using {{domxref("ReadableByteStreamController.enqueue()")}} (this is the only way to transfer data when using a "default" stream).

> The {{domxref("ReadableStreamBYOBRequest")}} has a {{domxref("ReadableStreamBYOBRequest.view","view")}} property, which is a view on the buffer allocated for transfer (the size of this buffer may be set using [`autoAllocateChunkSize`](/en-US/docs/Web/API/ReadableStream/ReadableStream#autoallocatechunksize) if desired).

Reviewer: > (the size of this buffer may be set using [`autoAllocateChunkSize`](/en-US/docs/Web/API/ReadableStream/ReadableStream#autoallocatechunksize) if desired)

This comment may be a little bit confusing, since if a BYOB reader is in use then the size is determined by the view passed to the BYOB reader.

Author: Good catch - I wrote this before the test code that verified how this works. 
Further down I talk about the autoAllocateChunkSize and get it right.

```suggestion
The {{domxref("ReadableStreamBYOBRequest")}} has a {{domxref("ReadableStreamBYOBRequest.view","view")}} property, which is a view on the buffer allocated for transfer.
```

      logConsumer(`Read ${bytesReceived} bytes`);
      result += value;

      // Add delay to emulate when data can't be read and data is enqueued
      if (bytesReceived > 300 && bytesReceived < 600) {
        logConsumer(`Delaying read to emulate slow stream reading`);
        const delay = (ms) => new Promise((resolve) => setTimeout(resolve, ms));
        await delay(1000);
      }

      // Read some more, and call this function again
      return reader.read( new Uint8Array(buffer, offset, buffer.byteLength - offset) ).then(processText);
    });
  }
}
```

#### Cancelling the stream using the reader

We can use {{domxref("ReadableStreamBYOBReader.cancel()")}} to cancel the stream.
For this example we call the method if a button is clicked with a reason "user choice" (other HTML and code for the button not shown).
We also log when the cancel operation completes.

```js
button.addEventListener('click', () => { reader.cancel("user choice").then( () => { logConsumer(`reader.cancel complete`) }) } );
```

{{domxref("ReadableStreamBYOBReader.releaseLock()")}} can be used to release the reader without cancelling the stream.

> Note however that this can only be called when there is no outsanding read request.

Author: Thanks. I'll say that, and add note to the other cases. 
```suggestion
Note however that any outstanding read requests will immediately be rejected.
A new reader can be acquired later on to read the remaining chunks.
```

This will also require something in the browser compatibility data. I'll think about that tomorrow.

> Note however that this can only be called when there is no outsanding read request.

Author: Ah, note that the releaseLock docs in this PR already fix this.. They say:

> If the reader's lock is released while it still has pending read requests then the promises returned by the reader's {{domxref("ReadableStreamBYOBReader.read()")}} method are immediately rejected with a `TypeError`.
Unread chunks remain in the stream's internal queue and can be read later by acquiring a new reader.

After calling `respond()` the `view` can no longer be written.

There is also an additional method {{domxref("ReadableStreamBYOBRequest.respondWithNewView()")}} to which an underlying source can pass a "new" view containing data to be transferred.
Note however that this new view must be over the _same_ memory buffer as the original, and from the same starting offset.
In other words, this method gives developers another way to send less than the full allocated buffer size (the other way being to specify the length in `respond()`).

Readable byte streams are normally read using a {{domxref("ReadableStreamBYOBReader")}}, which can be obtained by calling {{domxref("ReadableStream.getReader()")}} on the stream, specifying `mode: "byob"` in the options parameter.

A readable byte stream can also be read using a default reader ({{domxref("ReadableStreamDefaultReader")}}), but in this case `byobRequest` objects are only created when automatic buffer allocation is enabled for the stream ([`autoAllocateChunkSize`](/en-US/docs/Web/API/ReadableStream/ReadableStream#autoallocatechunksize) was set for the stream's `underlyingSource`).

> Note that the size indicated by `autoAllocateChunkSize` is used for the buffer size in this case; for a byte reader the size is set by the consumer.

Author: ```suggestion
Note that the size indicated by `autoAllocateChunkSize` is used for the buffer size in this case; for a byte reader the size is set by the buffer passed by the consumer.
```

---
title: Using readable byte streams
slug: Web/API/Streams_API/Using_readable_byte_streams

> tags:

Reviewer 2: ```suggestion
page-type: guide
tags:
```

> This is similar to the behaviour shown in the "pull: underlying byte source examples.

#### Mocked underlying socket source

The mocked underlying source has three important methods:

- `select2()` represents an outstanding request on the socket.
  It returns a promise that is resolved when data is available.
- `readInto()` reads data from the socket into a supplied buffer and then clears the data.
- `close()` closes the socket.

The implementation is very simplistic.
As shown below, `select2()` creates a randomly sized buffer of random data on a timeout.
The created data is read into a buffer then cleared in `readInto()`.

```js
class MockHypotheticalSocket {
  constructor() {
    this.max_data = 800; // total amount of data to stream from "socket"
    this.max_per_read = 100; // max data per read
    this.min_per_read = 40; // min data per read
    this.data_read = 0; // total data read so far (capped is maxdata)
    this.socketdata = null; // 
  }

  /* Method returning promise when this socket is readable. */
  select2() {
    // Object used to resolve promise
    const resultobj = {};
    resultobj["bytesRead"] = 0;

>     return new Promise((resolve, reject) => {

Reviewer 2: Am I correct that `reject` is not used here? Is it intended?

Author: You are correct. I could comment it out or remove it altogether. But I'm kind of OK with it as-is too.

Reviewer 2: No strong feeling, your call.

An alternative is:
```suggestion
    return new Promise((resolve/*, reject*/) => {
```

      if (this.data_read >= this.max_data) { //out of data
        resolve(resultobj);
        return;
      }

      // Emulate slow read of data
      window.setTimeout(() => {

>         const numberBytesRecieved = this.getNumberRandomBytesSocket();
>         this.data_read += numberBytesRecieved;
>         this.socketdata = this.randomByteArray(numberBytesRecieved);
>         resultobj["bytesRead"] = numberBytesRecieved;

Reviewer 2: ```suggestion
        const numberBytesReceived = this.getNumberRandomBytesSocket();
        this.data_read += numberBytesReceived;
        this.socketdata = this.randomByteArray(numberBytesReceived);
        resultobj["bytesRead"] = numberBytesReceived;
```

Author: Thanks!

#### Mocked underlying file source

For the underlying pull source we use the following class to (_very_ superficially) mock a nodejs [`FileHandle`](https://nodejs.org/api/fs.html#class-filehandle), and in particular the [`read()`](https://nodejs.org/api/fs.html#filehandlereadbuffer-offset-length-position) method.
The class generates random data to represent a file.
The `read()` method reads from this data into a provided buffer from the specified position.
The `close()` method does nothing: it is only provided to show where you might close the source when defining the constructor for the stream.

> **Note:** This same class is used for all the "pull source" examples.
> It is shown here for information only (so that it is obvious that it is a mock).

```js
class MockUnderlyingFileHandle {
  constructor() {
    this.maxdata = 1300; // "file size"
    this.filedata = this.randomByteArray(this.maxdata);
    this.position = 0;
  }

  /* Read data from "file" at position/length into specified buffer offset */
  read(buffer, offset, length, position) {
    // Object used to resolve promise
    const resultobj = {};
    resultobj["buffer"] = buffer;
    resultobj["bytesRead"] = 0;

>     return new Promise((resolve, reject) => {

Reviewer 2: `reject` not used?

Author: No. I guess it might be in some mocks though. We can fix if you like, but I'm inclined to leave it.

Author: ```suggestion
    return new Promise((resolve /*, reject*/) => {
```

      } else {
        position += bytesRead;
        controller.byobRequest.respond(bytesRead);
        logSource(`pull() with byobRequest. Transfer ${bytesRead} bytes`);
      }
    },
    cancel(reason) {
      // This is called if the stream is cancelled (via reader or controller).
      // Clean up any resources
      fileHandle.close();
      logSource(`cancel() with reason: ${reason}`);
    }
  });
}
```

#### Consuming the byte stream

The following code creates a `ReadableStreamBYOBReader` for the file byte stream and uses it read data into a buffer.
Note `processText()` is called recursively to read more data until the buffer is filled.
When the underlying source signals that it has no more data, the `reader.read()` will have `done` set to true, which in turn completes the read operation.

```js
const reader = stream.getReader({mode: "byob"});
let buffer = new ArrayBuffer(4000);
readStream(reader);

function readStream(reader) {
  let bytesReceived = 0;
  let offset =  0;

  while (offset < buffer.byteLength) {    
    // read() returns a promise that resolves when a value has been received

>     reader.read( new Uint8Array(buffer, offset, buffer.byteLength - offset) ).then(function processText({ done, value }) {

Reviewer 2: `processText`. Is this a common pattern? Naming a function as a parameter of the `then`, and "rescheduling" it in the `return` statement?

Author: It is the pattern used in all the examples I've seen for stream reader handling. Most of the examples out in the wild derive from the specification though, so it could just be that it is just the pattern most familiar to the spec authors.

Author: Yes, it works. Better for the default case than the bytestream reader though. Because in this case you have a fixed size buffer so you would in reality have to make sure you didn't run out of space. For the default you get handed a buffer rather than supplying your own, so you don't have to worry about that.

Not worth worrying about here.

```

#### Result

The logging from the underlying pull source (left) and consumer (right) are shown below.
Of particular note are that the:

- `start()` function is passed a `ReadableByteStreamController`
- the buffer passed to the reader is large enough to encompass the whole "file", so the whole file is transferred in one operation.

{{EmbedLiveSample("Underlying pull source","100%","500px")}}

### Underlying pull source with default reader

This live example shows how the same data might be read as a zero-copy transfer using a default reader ({{domxref("ReadableStreamDefaultReader")}}).
This uses the same same [mocked underlying file source](#mocked_underlying_file_source) as in the preceding example.

```js hidden
class MockUnderlyingFileHandle {
  constructor() {
    this.maxdata = 1300; // "file size"
    this.filedata = this.randomByteArray(this.maxdata);
    this.position = 0;
  }

  /* Read data from "file" at position/length into specified buffer offset */
  read(buffer, offset, length, position) {
    // Object used to resolve promise
    const resultobj = {};
    resultobj["buffer"] = buffer;
    resultobj["bytesRead"] = 0;

>     return new Promise((resolve, reject) => {

Reviewer 2: `reject` not used?

Author: Correct. The source for the mock is copied in a number of examples. This is hidden so it doesn't particularly matter.

Reviewer 2: No strong feeling, your call.

An alternative is:
```suggestion
    return new Promise((resolve/*, reject*/) => {
```

Author: Thanks @teoli2003. It's hard to know at the end of a big job like this whether you're being pragmatic or lazy. I'll make the change.

```js
button.addEventListener('click', () => { reader.cancel("user choice").then( () => { logConsumer(`reader.cancel complete`) }) } );
```

#### Result

The logging from the underlying bye pull source (left) and consumer (right) are shown below.

Note that the chunks are now 200-byte wide, as specified in the underlying byte source.
These are made as zero-copy transfers.

{{EmbedLiveSample("Underlying pull source with default reader","100%","500px")}}

### Underlying pull source with default reader and no allocation

For completeness, we can also use a default reader with a byte source that does not support automatic buffer allocation.

```js hidden
class MockUnderlyingFileHandle {
  constructor() {
    this.maxdata = 1300; // "file size"
    this.filedata = this.randomByteArray(this.maxdata);
    this.position = 0;
  }

  /* Read data from "file" at position/length into specified buffer offset */
  read(buffer, offset, length, position) {
    // Object used to resolve promise
    const resultobj = {};
    resultobj["buffer"] = buffer;
    resultobj["bytesRead"] = 0;

>     return new Promise((resolve, reject) => {

Reviewer 2: `reject` not used?

Reviewer 2: No strong feeling, your call.

An alternative is:
```suggestion
    return new Promise((resolve/*, reject*/) => {
```

This adds data to the stream's internal queues (this is the only way to add data to the stream in default streams).

The {{domxref("ReadableStreamBYOBRequest")}} has a {{domxref("ReadableStreamBYOBRequest.view","view")}} property, which is a view on the buffer allocated for transfer (the size of this buffer may be set using [`autoAllocateChunkSize`](/en-US/docs/Web/API/ReadableStream/ReadableStream#autoallocatechunksize) if desired).
Data from an underlying source should be written into this property, and then the underlying source must call {{domxref("ReadableStreamBYOBRequest.respond()","respond()")}} indicating the number of bytes written.
This signals that the data should be transferred, and the pending read request resolved.
After calling the `respond()` the `view` can no longer be written.

There is also an additional method {{domxref("ReadableStreamBYOBRequest.respondWithNewView()")}} to which an underlying source can pass a "new" view containing data to be transferred.
Note however that this new view must be over the _same_ memory buffer as the original, and from the same starting offset.

> In other words, this method gives developers another way to send less than the full allocated buffer size (the other way being to specify the length in `respond()`).

Author: @mgaudet You remember we discussed why you might use `respondWithNewView()`. My take on this is that it is just another way to provide a subset of the original buffer. We'll see what the spec authors say.

This doc will be the overview. Next come examples. They are coming together ... I think. You probably won't see any updates for two weeks as I am away. When I get back I'll get this to state for passing to spec authors. Any comments on what is there now are welcome though.

Author: Thanks very much @MattiasBuelens .

```suggestion
There is also an additional method {{domxref("ReadableStreamBYOBRequest.respondWithNewView()")}} to which an underlying source can pass a "new" view containing data to be transferred.
This new view must be over the _same_ memory buffer as the original, and from the same starting offset.
This method might be used if the underlying byte source needs to first transfer the view to a worker thread to populate (for example) and then get it back before responding to the `byobRequest`.
In most cases this method will not be needed.
```
